File: Problems/Algorithms/407. Trapping Rain Water II/TrapWaterII.kt
class Solution {
    fun trapRainWater(heightMap: Array<IntArray>): Int {
        val n = heightMap.size
        val m = heightMap[0].size
        
        val pq = PriorityQueue<IntArray>({ a, b -> a[2] - b[2] })
        val visited = Array(n) { BooleanArray(m) { false } }
        val neighbors = arrayOf(intArrayOf(0, 1), intArrayOf(0, -1), intArrayOf(1, 0), intArrayOf(-1, 0))
        
        for (i in 0..n-1) {
            pq.offer(intArrayOf(i, 0, heightMap[i][0]))
            pq.offer(intArrayOf(i, m-1, heightMap[i][m-1]))
            visited[i][0] = true
            visited[i][m-1] = true
        }
        
        for (j in 0..m-1) {
            pq.offer(intArrayOf(0, j, heightMap[0][j]))
            pq.offer(intArrayOf(n-1, j, heightMap[n-1][j]))
            visited[0][j] = true
            visited[n-1][j] = true
        }
        
        var ans = 0
        while (!pq.isEmpty()) {
            val curr = pq.poll()
            
            for (neighbor in neighbors) {
                val x = curr[0] + neighbor[0]
                val y = curr[1] + neighbor[1]
                if (x >= 0 && x < n && y >= 0 && y < m && !visited[x][y]) {
                    visited[x][y] = true
                    ans += maxOf(0, curr[2] - heightMap[x][y])
                    pq.offer(intArrayOf(x, y, maxOf(heightMap[x][y], curr[2])))
                }
            }
        }
        
        return ans
    }
}

File: Problems/Algorithms/407. Trapping Rain Water II/trap-water-II.cpp
#include <queue>
#include <vector>

using namespace std;

class Cell {
    public:
        int height;
        int row;
        int col;

        Cell(int height, int row, int col) {
            this->height = height;
            this->row = row;
            this->col = col;
        }

        bool operator<(const Cell& other) const {
            return height >= other.height;
        }
};

class Solution {
private:
    bool isValid(int row, int col, int rows, int cols) {
        return row >= 0 && col >= 0 && row < rows && col < cols;
    }

public:
    int trapRainWater(vector<vector<int>>& heightMap) {
        int directions[4][2] = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};

        int rows = heightMap.size(), cols = heightMap[0].size();

        vector<vector<bool>> visited(rows, vector<bool>(cols, false));
        priority_queue<Cell> boundary;

        for (int i = 0; i < rows; i++) {
            boundary.push(Cell(heightMap[i][0], i, 0));
            boundary.push(Cell(heightMap[i][cols-1], i, cols-1));
            visited[i][0] = visited[i][cols-1] = true;
        }

        for (int j = 0; j < cols; j++) {
            boundary.push(Cell(heightMap[0][j], 0, j));
            boundary.push(Cell(heightMap[rows-1][j], rows-1, j));
            visited[0][j] = visited[rows-1][j] = true;
        }

        int volume = 0;
        while (!boundary.empty()) {
            Cell curr = boundary.top();
            boundary.pop();

            int row = curr.row;
            int col = curr.col;
            int height = curr.height;

            for (int direction = 0; direction < 4; direction++) {
                int nextRow = row + directions[direction][0];
                int nextCol = col + directions[direction][1];

                if (isValid(nextRow, nextCol, rows, cols) && !visited[nextRow][nextCol]) {
                    int nextHeight = heightMap[nextRow][nextCol];

                    if (nextHeight < height) {
                        volume += height - nextHeight;
                    }

                    boundary.push(Cell(max(nextHeight, height), nextRow, nextCol));
                    visited[nextRow][nextCol] = true;
                }
            }
        }

        return volume;
    }
};
